src/v/archival/archival_policy.h

```diff
@@ -32,10 +40,30 @@ struct upload_candidate {
     size_t final_file_offset;
     model::timestamp base_timestamp;
     model::timestamp max_timestamp;
+    std::vector<ss::lw_shared_ptr<storage::segment>> sources;
```

Comment:
I created a struct which contains upload candidate and the locks `upload_candidate_with_locks`, because if we keep locks inside the upload candidate then copying the candidate is prohibited, and we copy it in a couple of places when scheduling the upload (eg when uploading segment and transaction manifest). 

The locks are created when the candidate is created and then later moved into the scheduled upload.